berbix-java/src/main/java/com/berbix/BerbixApi.java

```diff
@@ -139,153 +127,201 @@ private Tokens createTokens(FetchTokensResponse fetchTokensResponse) {
-    refreshIfNecessary(tokens);
-    Request request = jettyClient.newRequest(apiHost + path);
-    request.method(method);
-    request.header("Authorization", "Bearer " + tokens.accessToken);
-    request.header("Content-Type", "application/json");
-    request.header("Accept", "application/json");
-    request.header("User-Agent", "BerbixJava/" + Berbix.BERBIX_SDK_VERSION);
-
-    if (payload != null) {
-      BytesContentProvider bytesRequestContent = new BytesContentProvider("application/json",
-          objectMapper.writeValueAsString(payload).getBytes(StandardCharsets.UTF_8));
-      request.content(bytesRequestContent, "application/json");
-    }
+  private <T> CompletableFuture<T> tokenAuthRequest(String method, Tokens tokens, String path, Object payload, Class<T> responseClass) throws IOException {
+    return refreshIfNecessaryAsync(tokens).thenCompose(newTokens -> {
+      Request request = jettyClient.newRequest(apiHost + path);
+      request.method(method);
+      request.header("Authorization", "Bearer " + newTokens.accessToken);
+      request.header("Content-Type", "application/json");
+      request.header("Accept", "application/json");
+      request.header("User-Agent", "BerbixJava/" + Berbix.BERBIX_SDK_VERSION);
+
+      if (payload != null) {
+        BytesContentProvider bytesRequestContent;
+        try {
+          bytesRequestContent = new BytesContentProvider("application/json",
+              objectMapper.writeValueAsString(payload).getBytes(StandardCharsets.UTF_8));
+        } catch (JsonProcessingException ex) {
+          throw new BerbixException("Unable to serialize API request", ex);
+        }
+        request.content(bytesRequestContent, "application/json");
+      }
 
-    ContentResponse contentResponse;
-    try {
+      CompletableFuture<ContentResponse> completableFuture = new CompletableFuture<>();
       request.timeout(REQUEST_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
-      contentResponse = request.send();
-      if (contentResponse.getStatus() == HttpStatus.OK_200) {
-
-        // uncomment to see raw response
-        // System.out.println("response:" + response);
-
-        return contentResponse.getContentAsString();
-      } else if (contentResponse.getStatus() == 204) {
-        return "finished";
-      } else {
-        String response = contentResponse.getContentAsString();
-        throw new IOException(response);
-      }
-    } catch (InterruptedException | TimeoutException | ExecutionException e) {
-      // e.printStackTrace();
-      throw new IOException(e.getMessage());
-    }
+      request.send(new CompletableFutureResponseListener(completableFuture));
+
+      return completableFuture.thenApply(response -> {
+        if (response.getStatus() == 204 && responseClass == String.class) {
+          // cast string as String so T compiles.
+          return responseClass.cast("finished");
+        } else {
+          String responseData = response.getContentAsString();
```

Comment:
How does a caller of public methods that call `tokenAuthRequest()` detect if a non-2XX code was returned e.g. if they have a bad API secret?

Reply:
They will get an exception they need to handle from. https://github.com/berbix/berbix-java/pull/2/files#diff-2cee81cb9fe60f33dcf4e0de932378d6639372e2a83b6184ba3c42888f202567R315

Reply:
Can you point me at the docs for what status codes are considered failures? I skimmed the documentation, but I couldn't figure out what exactly constitutes "failures"

Reply:
It just checks whether there is a request exception or a response exception. https://www.eclipse.org/jetty/javadoc/jetty-9/org/eclipse/jetty/client/api/Result.html#isFailed()

From what I could tell, 4xx and 5xx both were considered `failures` and 2xx was considered success. I didn't have any 3xx, but I imagine they would be failures too.

Reply:
Got it, mind leaving a comment somewhere with you findings?

Reply:
done